Sources/Utility/Process.swift

```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
         self.redirectOutput = redirectOutput 
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    internal func validateArgumentIsExecutable(_ program: String) throws {
+
+        // check to see if the argument is already listed in the vaildated cache
+        guard !Process.validatedExecutables.contains(program) else {
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
+        guard let executableLocation = Utility.lookupExecutablePath(filename: program, searchPaths: envSearchPaths) else {
+            throw Process.Error.missingProgram(program: program)
+        }
+        guard localFileSystem.isExecutableFile(executableLocation) else {
+            throw Process.Error.programNotExecutable(program: program)
+        }
+        // add the program name to the validated cache
+        Process.validatedExecutables.insert(program)
```

Comment:
I'm not entirely sure what you mean by "store the result", given that line 190 is appending the name of the program searched for, and found, in the Set. Do you mean store the result of where we found the executable? Even though we don't pass that back or use it? Or is that something we should be doing?

I thought about the thread-safe, and worst case that I could spot was that extra checking would happen, and the Set might be updated with an additional redundant value (and subsequently ignore it I presume). I'm feeling a little bait-and-switched with this implementation though, because my original proposal wouldn't have had to concern itself so much with thread safety since it was very early and simply in the execution path and before the system would start to attempt to do "interesting things" in parallel for efficiency. So - that said - I'm also not immediately clueful about *how* to make this thread safe, could you perhaps suggest a specific path to enabling that? Are you thinking that we should wrap the shared state (the cache) in a lock/protection akin to what is described at https://www.bignerdranch.com/blog/locking-in-swift-helping-protect-me-from-myself/ or some of the answers at http://stackoverflow.com/questions/28191079/create-thread-safe-array-in-swift?